python/cudf/cudf/testing/testing.py

```diff
@@ -201,31 +206,49 @@ def assert_column_equal(
     ):
         left = left.astype(left.categories.dtype)
         right = right.astype(right.categories.dtype)
-
     columns_equal = False
-    try:
-        columns_equal = (
-            (
-                cp.all(left.isnull().values == right.isnull().values)
-                and cp.allclose(
+    if left.size == right.size == 0:
+        columns_equal = True
+    elif not (
+        (is_string_dtype(left) and is_numeric_dtype(right))
+        or (is_numeric_dtype(left) and is_string_dtype(right))
+    ):
+        try:
+            # nulls must be in the same places for all dtypes
+            nulls_equal = cp.all(left.isnull().values == right.isnull().values)
+
+            if not check_exact and is_numeric_dtype(left):
+                # non-null values must be the same
+                values_equal = cp.allclose(
                     left[left.isnull().unary_operator("not")].values,
                     right[right.isnull().unary_operator("not")].values,
                 )
-            )
-            if not check_exact and is_numeric_dtype(left)
-            else left.equals(right)
-        )
-    except TypeError as e:
-        if str(e) != "Categoricals can only compare with the same type":
-            raise e
-        if is_categorical_dtype(left) and is_categorical_dtype(right):
-            left = left.astype(left.categories.dtype)
-            right = right.astype(right.categories.dtype)
+                if not left.dtype.kind == right.dtype.kind == "f":
+                    columns_equal = nulls_equal and values_equal
+                else:
+                    # nans must be the same for float types
+                    nans_equal = cp.all(
+                        is_nan(left).values == is_nan(right).values
+                    )
+                    columns_equal = nulls_equal and values_equal and nans_equal
+            else:
+                columns_equal = left.equals(right)
+        except TypeError as e:
+            if str(e) != "Categoricals can only compare with the same type":
+                raise e
+            if is_categorical_dtype(left) and is_categorical_dtype(right):
+                left = left.astype(left.categories.dtype)
+                right = right.astype(right.categories.dtype)
     if not columns_equal:
-        msg1 = f"{left.values_host}"
-        msg2 = f"{right.values_host}"
+        ldata = [val for val in left.to_pandas(nullable=True)]
+        rdata = [val for val in right.to_pandas(nullable=True)]
+        msg1 = f"{ldata}"
+        msg2 = f"{rdata}"
```

Comment:
These can just be inlined into the `raise_assert_detail` function call. Pass by kwarg if you think that's helpful in documenting the purpose of these messages.